src/Domain/LeanCode.DomainModels.MassTransitRelay/Outbox/NewtonsoftJsonEventsSerializer.cs

```diff
@@ -0,0 +1,69 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Reflection;
+using LeanCode.Components;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace LeanCode.DomainModels.MassTransitRelay.Outbox
+{
+    public class NewtonsoftJsonEventsSerializer : IRaisedEventsSerializer
+    {
+        private readonly TypesCatalog typesCatalog;
+        private readonly ConcurrentDictionary<string, Type> deserializerCache = new();
+        private readonly JsonSerializerSettings settings;
+
+        public NewtonsoftJsonEventsSerializer(TypesCatalog typesCatalog)
+            : this(typesCatalog, null)
+        { }
+
+        public NewtonsoftJsonEventsSerializer(TypesCatalog typesCatalog, JsonSerializerSettings? settings)
+        {
+            this.typesCatalog = typesCatalog;
+            this.settings = settings ?? new JsonSerializerSettings
+            {
+                ContractResolver = new ContractResolver(),
+            };
+        }
+
+        public RaisedEvent WrapEvent(object evt, Guid correlationId)
+        {
+            return RaisedEvent.Create(evt, correlationId, JsonConvert.SerializeObject(evt, settings));
+        }
+
+        public object ExtractEvent(RaisedEvent evt)
+        {
+            var type = deserializerCache.GetOrAdd(evt.EventType, GetEventType);
+            return JsonConvert.DeserializeObject(evt.Payload, type, settings)!;
+        }
+
+        private Type GetEventType(string type)
+        {
+            return typesCatalog.Assemblies
+                .Select(ass => ass.GetType(type))
+                .Where(t => t != null)
+                .FirstOrDefault()
+                ?? throw new InvalidOperationException($"Type: {type} is not defined in any of contract assemblies");
```

Comment:
Just `First` would be descriptive enough I think.

Reply:
Considering that it might be thrown in apps and it will be raised from the depths of processing pipeline, I think that as much additional info as possible is better. `First` will throw a very generic error and sometimes stack trace might not be available (logs).